### ISSUE TITLE
[DX] API documentation: mention that it's not necessary to delete own config on uninstall

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#5049]
We could add documentation stating why it's not necessary to delete config files in `hook_uninstall()`. I think @indigoxela said it best:

> As long as a module properly implements `hook_config_info()`, it's not necessary to delete this config on uninstall. If, however, a module writes to config files provided by other modules, it should remove its own values from those other files.

---

Original issue:

**Description of the bug**

The https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_uninstall/1 pages gives this example code:

```
function hook_uninstall() {
  state_del('my_module_last_cron');
}
```
However, while clearing specific states or variables is a good practice upon a module uninstall, the configuration files must also be deleted. While https://github.com/backdrop/backdrop-issues/issues/4189 is about enabling core to cleanup after lazy modules that didn't delete their configuration files, this one is different in that I suggest to slightly change the API doc adding one more line like:

```
function hook_uninstall() {
  state_del('my_module_last_cron');
  config('my_module.settings')->delete();
}
```

which would make it easier for all Backdrop developers, especially new ones, copying the function from the https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_uninstall/1 page to also grab the deletion line, (instead of additionally searching on how to automatically delete the configuration file - nobody does this). One extra line doesn't hurt, but it can help to keep lot's of Backdrop sites cleaner.

**Comment by klonos** [comment]
Thanks @alanmels ...I initially RTBC'ed this, but can we please also add something in this bit of the docblock of the `hook_uninstall()` function:

```php
 * The information that the module should remove includes:
 * - settings that the module has set using state_set().
 * - modifications to existing tables
```

...we should add a bullet point:

```php
 * The information that the module should remove includes:
 * - settings that the module has set using state_set()
 * - modifications to existing tables
 * - any configuration files related to the module <-- ADD THIS
```

**Comment by indigoxela** [comment]
@alanmels As long as a module properly implements [hook_config_info()](https://docs.backdropcms.org/api/backdrop/core%21modules%21config%21config.api.php/function/hook_config_info/1), it's not necessary to delete this config on uninstall. If, however, modules write to config files provided by other modules, they should do the cleanup.

If a module provides config files, but does not implement hook_config_info(), there will be errors on configuration sync, so that hook is essential.

**Comment by indigoxela** [comment]
I'd suggest to use a different example for config cleanup, because `my_module.settings` is usually something that ships with a hook_config_info().

**Comment by alanmels** [comment]
@klonos, for some reason I can't edit the file anymore:
 
![Screenshot from 2021-04-20 09-04-26](https://user-images.githubusercontent.com/56604405/115419291-72ad9e00-a1b7-11eb-9e29-fea44a2fd7e4.png)

Also with @indigoxela suggestion I'm not sure what's the ideal resolution to this. I just want the docs embrace all scenarios and make it easy for the developers to find cleanup functions.

**Comment by klonos** [comment]
> @klonos, for some reason I can't edit the file anymore:

🤔 ...I don't think that I have anything to do with that 🤷🏼‍♂️ 

> Also with @indigoxela suggestion I'm not sure what's the ideal resolution to this....

I have an idea, just no time ATM. I'll come back to this.

**Comment by jenlampton** [comment]
I also agree that we shouldn't include unnecessary steps in the documentation or in example code. That would be more likely to confuse people than to help. Instead, we could add documentation stating why it's not necessary. I think @indigoxela said it best:

> As long as a module properly implements `hook_config_info()`, it's not necessary to delete this config on uninstall. If, however, a module writes to config files provided by other modules, it should remove its own values from those other files.

Can we include the text above somewhere in there?
